host/xtest/pkcs11_1000.c

```diff
@@ -3741,3 +3742,863 @@ static void xtest_pkcs11_test_1016(ADBG_Case_t *c)
+		.data_size = sizeof(digest_test_pattern) - 1,
+		.digest = digest_test_pattern_sha256,
+		.digest_size = sizeof(digest_test_pattern_sha256),
+	},
+	{
+		.test_name = "CKM_SHA384/empty",
+		.mecha = CKM_SHA384,
+		.data = digest_test_pattern_empty,
+		.data_size = sizeof(digest_test_pattern_empty) - 1,
+		.digest = digest_test_pattern_empty_sha384,
+		.digest_size = sizeof(digest_test_pattern_empty_sha384),
+	},
+	{
+		.test_name = "CKM_SHA384/test pattern",
+		.mecha = CKM_SHA384,
+		.data = digest_test_pattern,
+		.data_size = sizeof(digest_test_pattern) - 1,
+		.digest = digest_test_pattern_sha384,
+		.digest_size = sizeof(digest_test_pattern_sha384),
+	},
+	{
+		.test_name = "CKM_SHA512/empty",
+		.mecha = CKM_SHA512,
+		.data = digest_test_pattern_empty,
+		.data_size = sizeof(digest_test_pattern_empty) - 1,
+		.digest = digest_test_pattern_empty_sha512,
+		.digest_size = sizeof(digest_test_pattern_empty_sha512),
+	},
+	{
+		.test_name = "CKM_SHA512/test pattern",
+		.mecha = CKM_SHA512,
+		.data = digest_test_pattern,
+		.data_size = sizeof(digest_test_pattern) - 1,
+		.digest = digest_test_pattern_sha512,
+		.digest_size = sizeof(digest_test_pattern_sha512),
+	},
+};
+
+static CK_ATTRIBUTE digest_generate_aes_object[] = {
+	{ CKA_CLASS, &(CK_OBJECT_CLASS) { CKO_SECRET_KEY },
+	  sizeof(CK_OBJECT_CLASS) },
+	{ CKA_KEY_TYPE, &(CK_KEY_TYPE) { CKK_AES }, sizeof(CK_KEY_TYPE) },
+	{ CKA_TOKEN, &(CK_BBOOL) { CK_FALSE }, sizeof(CK_BBOOL) },
+	{ CKA_PRIVATE, &(CK_BBOOL) { CK_FALSE }, sizeof(CK_BBOOL) },
+	{ CKA_SENSITIVE, &(CK_BBOOL) { CK_FALSE }, sizeof(CK_BBOOL) },
+	{ CKA_EXTRACTABLE, &(CK_BBOOL) { CK_TRUE }, sizeof(CK_BBOOL) },
+	{ CKA_VALUE_LEN, &(CK_ULONG) { 16 }, sizeof(CK_ULONG) },
+};
+
+static CK_ATTRIBUTE digest_generate_gensecret_object[] = {
+	{ CKA_CLASS, &(CK_OBJECT_CLASS) { CKO_SECRET_KEY },
+	  sizeof(CK_OBJECT_CLASS) },
+	{ CKA_KEY_TYPE, &(CK_KEY_TYPE) { CKK_GENERIC_SECRET },
+	  sizeof(CK_KEY_TYPE) },
+	{ CKA_TOKEN, &(CK_BBOOL) { CK_FALSE }, sizeof(CK_BBOOL) },
+	{ CKA_PRIVATE, &(CK_BBOOL) { CK_FALSE }, sizeof(CK_BBOOL) },
+	{ CKA_SENSITIVE, &(CK_BBOOL) { CK_FALSE }, sizeof(CK_BBOOL) },
+	{ CKA_EXTRACTABLE, &(CK_BBOOL) { CK_TRUE }, sizeof(CK_BBOOL) },
+	{ CKA_VALUE_LEN, &(CK_ULONG) { 16 }, sizeof(CK_ULONG) },
+};
+
+static CK_ATTRIBUTE digest_data_object[] = {
+	{ CKA_CLASS, &(CK_OBJECT_CLASS) { CKO_DATA },
+	  sizeof(CK_OBJECT_CLASS) },
+	{ CKA_TOKEN, &(CK_BBOOL) { CK_FALSE }, sizeof(CK_BBOOL) },
+	{ CKA_PRIVATE, &(CK_BBOOL) { CK_FALSE }, sizeof(CK_BBOOL) },
+};
+
+static void xtest_pkcs11_test_1020(ADBG_Case_t *c)
+{
+	CK_RV rv = CKR_GENERAL_ERROR;
+	CK_SLOT_ID slot = 0;
+	CK_SESSION_HANDLE session = CK_INVALID_HANDLE;
+	CK_FLAGS session_flags = CKF_SERIAL_SESSION | CKF_RW_SESSION;
+	bool logged_in = false;
+	uint8_t data[128] = { 0 };
+	CK_ULONG data_size = 0;
+	uint8_t digest[64] = { 0 };
+	CK_ULONG digest_size = 0;
+	const uint8_t *expect_digest = NULL;
+	CK_ULONG expect_digest_size = 0;
+	CK_MECHANISM mechanism = {
```

Comment:
i think `CK_MECHANISM mechanism = { CKM_MD5, NULL, 0 };` is fine.

Reply:
OK